### Multiple Linear Regression

Each problem is wirth 20 points.

### Problem 1
Denote the data set with all the columns except "Column 0" and "rate" as $x$. Then denote the "rate" variable with $y$. Train a Linear Regression model with $y$ as a dependent variable and $x$ as a set of independent variables. Make sure to add constant-coefficient in $x$ before running the model. Report the summary, $R^2$ and all the variables with the p-value of  $p>0.10$.

In [78]:
import pandas as pd
import numpy as np
import matplotlib as mpt
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
data = pd.read_csv('Highway1-2.csv')
data.columns

Index(['Unnamed: 0', 'rate', 'len', 'ADT', 'trks', 'sigs1', 'slim', 'shld',
       'lane', 'acpt', 'itg', 'lwid', 'hwy'],
      dtype='object')

In [79]:
y = data['rate']

In [80]:
x = data[['len', 'ADT', 'trks', 'sigs1', 'slim', 'shld', 'lane', 'acpt', 'itg', 'lwid']]
x = sm.add_constant(x)

In [81]:
data.head()

Unnamed: 0  rate    len  ADT  trks     sigs1  slim  shld  lane  acpt   itg  \
0           1  4.58   4.99   69     8  0.200401    55    10     8   4.6  1.20   
1           2  2.86  16.11   73     8  0.062073    60    10     4   4.4  1.43   
2           3  3.02   9.75   49    10  0.102564    60    10     4   4.7  1.54   
3           4  2.29  10.65   61    13  0.093897    65    10     6   3.8  0.94   
4           5  1.61  20.01   28    12  0.049975    70    10     4   2.2  0.65   

   lwid  hwy  
0    12  FAI  
1    12  FAI  
2    12  FAI  
3    12  FAI  
4    12  FAI

In [82]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.743
Model:                            OLS   Adj. R-squared:                  0.651
Method:                 Least Squares   F-statistic:                     8.080
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           5.72e-06
Time:                        16:12:16   Log-Likelihood:                -55.125
No. Observations:                  39   AIC:                             132.2
Df Residuals:                      28   BIC:                             150.5
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.8175      5.969      2.147      0.041       0.591      25.044
len           -0.0607      0.033     -1.842      0.076      -0.128       0.007
ADT            0.0037      0.032      0.114      0.910      -0.062       0.069
trks          -0.1148      0.103     -1.119      0.272      -0.325       0.095
sigs1          0.4053      0.404      1.002      0.325      -0.423       1.233
slim          -0.0593      0.063     -0.947      0.352      -0.188       0.069
shld          -0.0908      0.108     -0.844      0.406      -0.311       0.130
lane           0.0197      0.274      0.072      0.943      -0.542       0.582
acpt           0.0885      0.031      2.858      0.008       0.025       0.152
itg            0.2059      1.135      0.181      0.857      -2.119       2.531
lwid          -0.3867      0.475     -0.815      0.422      -1.359       0.585
==============================================================================
Omnibus:                        0.400   Durbin-Watson:                   2.537
Prob(Omnibus):                  0.819   Jarque-Bera (JB):                0.239
Skew:                           0.187   Prob(JB):                        0.887
Kurtosis:                       2.913   Cond. No.                     2.03e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.03e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R^2 is 0.743. Variables with P > 0.9 are ADT and lane.

### Problem 2
In the data set, find out which 3 variables have the highest positive correlation with the rate variable and which 3 have the highest negative correlation? Now check the p-values of these 6 variables in a summary of a model from problem 1. Explain what do you observe.

In [83]:
data = data.drop('Unnamed: 0', 1)

In [84]:
data.corr()

rate       len       ADT      trks     sigs1      slim      shld  \
rate   1.000000 -0.465290 -0.028570 -0.512522  0.603191 -0.680984 -0.386907   
len   -0.465290  1.000000 -0.271569  0.495943 -0.391851  0.186243 -0.104926   
ADT   -0.028570 -0.271569  1.000000 -0.096682  0.159381  0.244157  0.457307   
trks  -0.512522  0.495943 -0.096682  1.000000 -0.475308  0.296184  0.006135   
sigs1  0.603191 -0.391851  0.159381 -0.475308  1.000000 -0.424415 -0.124103   
slim  -0.680984  0.186243  0.244157  0.296184 -0.424415  1.000000  0.689009   
shld  -0.386907 -0.104926  0.457307  0.006135 -0.124103  0.689009  1.000000   
lane  -0.032979 -0.202504  0.823930 -0.153324  0.263967  0.264520  0.481771   
acpt   0.752025 -0.238706 -0.223980 -0.360266  0.513461 -0.681521 -0.424951   
itg   -0.024841 -0.247562  0.903701 -0.067231  0.087017  0.241282  0.375022   
lwid  -0.005619 -0.310650  0.127878 -0.155271  0.062084  0.098693 -0.042896   

           lane      acpt       itg      lwid  
rate  -0.032979  0.752025 -0.024841 -0.005619  
len   -0.202504 -0.238706 -0.247562 -0.310650  
ADT    0.823930 -0.223980  0.903701  0.127878  
trks  -0.153324 -0.360266 -0.067231 -0.155271  
sigs1  0.263967  0.513461  0.087017  0.062084  
slim   0.264520 -0.681521  0.241282  0.098693  
shld   0.481771 -0.424951  0.375022 -0.042896  
lane   1.000000 -0.208779  0.697913  0.095723  
acpt  -0.208779  1.000000 -0.200158 -0.042013  
itg    0.697913 -0.200158  1.000000  0.102883  
lwid   0.095723 -0.042013  0.102883  1.000000

Variables with the highest positive correlation with rate are acpt and sigs1

Variables with the highest negative correlation with rate are slim, trks, and len

The p-values of acpt and sigs1 are 0.008 and 0.325. The p-values of slim, trks, and len are 0.352, 0.272, and 0.076. There is no relation beyween p-value and correlation.

### Problem 3
Find and remove a variable that has the highest p-vlaue from the list $x$ in the summary of the model you created in problem 1. Now create a new model (y is still the dependent variable) with this new list of variables and display the summary. Repeat this process by picking one variable with highest p-value at a time and making a new model until you run out of the variable with a p-value of more than $0.10$ in the summary. 
Report the summary of the final model together with $R^2$.

Variable with the highest p-value is lane.

In [85]:
x = data[['len', 'ADT', 'trks', 'sigs1', 'slim', 'shld', 'acpt', 'itg', 'lwid']]
x = sm.add_constant(x)

In [86]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.743
Model:                            OLS   Adj. R-squared:                  0.663
Method:                 Least Squares   F-statistic:                     9.296
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.79e-06
Time:                        16:12:16   Log-Likelihood:                -55.128
No. Observations:                  39   AIC:                             130.3
Df Residuals:                      29   BIC:                             146.9
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.8342      5.861      2.190      0.037       0.847      24.822
len           -0.0603      0.032     -1.884      0.070      -0.126       0.005
ADT            0.0049      0.026      0.187      0.853      -0.049       0.059
trks          -0.1156      0.100     -1.153      0.258      -0.321       0.089
sigs1          0.4142      0.378      1.096      0.282      -0.359       1.187
slim          -0.0590      0.061     -0.961      0.344      -0.185       0.067
shld          -0.0899      0.105     -0.856      0.399      -0.305       0.125
acpt           0.0883      0.030      2.914      0.007       0.026       0.150
itg            0.1955      1.106      0.177      0.861      -2.067       2.458
lwid          -0.3866      0.466     -0.829      0.414      -1.340       0.567
==============================================================================
Omnibus:                        0.374   Durbin-Watson:                   2.544
Prob(Omnibus):                  0.830   Jarque-Bera (JB):                0.244
Skew:                           0.186   Prob(JB):                        0.885
Kurtosis:                       2.890   Cond. No.                     2.03e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.03e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [87]:
x = data[['len', 'ADT', 'trks', 'sigs1', 'slim', 'shld', 'acpt', 'lwid']]
x = sm.add_constant(x)

In [88]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.742
Model:                            OLS   Adj. R-squared:                  0.674
Method:                 Least Squares   F-statistic:                     10.80
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           5.23e-07
Time:                        16:12:16   Log-Likelihood:                -55.149
No. Observations:                  39   AIC:                             128.3
Df Residuals:                      30   BIC:                             143.3
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.8462      5.765      2.228      0.034       1.072      24.621
len           -0.0609      0.031     -1.945      0.061      -0.125       0.003
ADT            0.0091      0.012      0.751      0.458      -0.016       0.034
trks          -0.1159      0.099     -1.176      0.249      -0.317       0.085
sigs1          0.4040      0.368      1.099      0.280      -0.347       1.155
slim          -0.0571      0.059     -0.961      0.344      -0.179       0.064
shld          -0.0938      0.101     -0.929      0.360      -0.300       0.112
acpt           0.0889      0.030      3.003      0.005       0.028       0.149
lwid          -0.3955      0.456     -0.867      0.393      -1.327       0.536
==============================================================================
Omnibus:                        0.446   Durbin-Watson:                   2.555
Prob(Omnibus):                  0.800   Jarque-Bera (JB):                0.272
Skew:                           0.200   Prob(JB):                        0.873
Kurtosis:                       2.915   Cond. No.                     2.03e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.03e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [89]:
x = data[['len', 'trks', 'sigs1', 'slim', 'shld', 'acpt', 'lwid']]
x = sm.add_constant(x)

In [90]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.737
Model:                            OLS   Adj. R-squared:                  0.678
Method:                 Least Squares   F-statistic:                     12.44
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.82e-07
Time:                        16:12:16   Log-Likelihood:                -55.513
No. Observations:                  39   AIC:                             127.0
Df Residuals:                      31   BIC:                             140.3
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.6986      5.721      2.219      0.034       1.030      24.367
len           -0.0645      0.031     -2.098      0.044      -0.127      -0.002
trks          -0.1114      0.098     -1.140      0.263      -0.311       0.088
sigs1          0.4680      0.355      1.319      0.197      -0.256       1.192
slim          -0.0615      0.059     -1.046      0.304      -0.181       0.058
shld          -0.0680      0.094     -0.721      0.476      -0.260       0.124
acpt           0.0840      0.029      2.929      0.006       0.026       0.142
lwid          -0.3603      0.450     -0.800      0.430      -1.279       0.558
==============================================================================
Omnibus:                        0.122   Durbin-Watson:                   2.419
Prob(Omnibus):                  0.941   Jarque-Bera (JB):                0.229
Skew:                           0.120   Prob(JB):                        0.892
Kurtosis:                       2.711   Cond. No.                     1.92e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.92e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [91]:
x = data[['len', 'trks', 'sigs1', 'slim', 'acpt', 'lwid']]
x = sm.add_constant(x)

In [92]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.733
Model:                            OLS   Adj. R-squared:                  0.683
Method:                 Least Squares   F-statistic:                     14.65
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           5.70e-08
Time:                        16:12:16   Log-Likelihood:                -55.837
No. Observations:                  39   AIC:                             125.7
Df Residuals:                      32   BIC:                             137.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.4384      5.667      2.195      0.036       0.895      23.982
len           -0.0581      0.029     -1.989      0.055      -0.118       0.001
trks          -0.1014      0.096     -1.057      0.299      -0.297       0.094
sigs1          0.4357      0.349      1.247      0.222      -0.276       1.148
slim          -0.0901      0.043     -2.095      0.044      -0.178      -0.003
acpt           0.0847      0.028      2.979      0.005       0.027       0.143
lwid          -0.2600      0.425     -0.611      0.545      -1.126       0.606
==============================================================================
Omnibus:                        0.152   Durbin-Watson:                   2.360
Prob(Omnibus):                  0.927   Jarque-Bera (JB):                0.262
Skew:                           0.134   Prob(JB):                        0.877
Kurtosis:                       2.701   Cond. No.                     1.90e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.9e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [93]:
x = data[['len', 'trks', 'sigs1', 'slim', 'acpt']]
x = sm.add_constant(x)

In [94]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.730
Model:                            OLS   Adj. R-squared:                  0.689
Method:                 Least Squares   F-statistic:                     17.84
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           1.51e-08
Time:                        16:12:16   Log-Likelihood:                -56.064
No. Observations:                  39   AIC:                             124.1
Df Residuals:                      33   BIC:                             134.1
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.4004      2.700      3.482      0.001       3.907      14.894
len           -0.0531      0.028     -1.912      0.065      -0.110       0.003
trks          -0.0986      0.095     -1.038      0.307      -0.292       0.095
sigs1          0.4378      0.346      1.265      0.215      -0.266       1.142
slim          -0.0931      0.042     -2.201      0.035      -0.179      -0.007
acpt           0.0851      0.028      3.022      0.005       0.028       0.142
==============================================================================
Omnibus:                        0.378   Durbin-Watson:                   2.337
Prob(Omnibus):                  0.828   Jarque-Bera (JB):                0.510
Skew:                           0.198   Prob(JB):                        0.775
Kurtosis:                       2.603   Cond. No.                         895.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [95]:
x = data[['len', 'sigs1', 'slim', 'acpt']]
x = sm.add_constant(x)

In [96]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.721
Model:                            OLS   Adj. R-squared:                  0.688
Method:                 Least Squares   F-statistic:                     21.98
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           4.95e-09
Time:                        16:12:16   Log-Likelihood:                -56.691
No. Observations:                  39   AIC:                             123.4
Df Residuals:                      34   BIC:                             131.7
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          8.6349      2.600      3.321      0.002       3.350      13.919
len           -0.0641      0.026     -2.492      0.018      -0.116      -0.012
sigs1          0.5280      0.335      1.574      0.125      -0.154       1.210
slim          -0.0948      0.042     -2.240      0.032      -0.181      -0.009
acpt           0.0879      0.028      3.133      0.004       0.031       0.145
==============================================================================
Omnibus:                        0.854   Durbin-Watson:                   2.336
Prob(Omnibus):                  0.653   Jarque-Bera (JB):                0.904
Skew:                           0.307   Prob(JB):                        0.636
Kurtosis:                       2.576   Cond. No.                         850.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [97]:
x = data[['len', 'slim', 'acpt']]
x = sm.add_constant(x)

In [98]:
model = sm.OLS(y,x)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.701
Model:                            OLS   Adj. R-squared:                  0.675
Method:                 Least Squares   F-statistic:                     27.33
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           2.76e-09
Time:                        16:12:16   Log-Likelihood:                -58.062
No. Observations:                  39   AIC:                             124.1
Df Residuals:                      35   BIC:                             130.8
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.3256      2.617      3.564      0.001       4.014      14.638
len           -0.0771      0.025     -3.102      0.004      -0.128      -0.027
slim          -0.1024      0.043     -2.386      0.023      -0.190      -0.015
acpt           0.1014      0.027      3.721      0.001       0.046       0.157
==============================================================================
Omnibus:                        0.311   Durbin-Watson:                   2.155
Prob(Omnibus):                  0.856   Jarque-Bera (JB):                0.493
Skew:                           0.097   Prob(JB):                        0.782
Kurtosis:                       2.485   Cond. No.                         837.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R^2 is 0.701

### Problem 4
Filter the original data set only with the significant variable you found in problem 3 and the rate variable. Divide the data set into training and testing with 80 % by 20% rule. Train a new model using training data, report the summary and $R^2$.

In [99]:
from sklearn.model_selection import train_test_split

In [100]:
xx = data[['len', 'slim', 'acpt']]
y = data.rate
xtrain, xtest, ytrain, ytest = model_selection.train_test_split(xx, y, test_size = 0.2, random_state = 17)

In [101]:
modelxx = sm.OLS(ytrain, sm.add_constant(xtrain)).fit()
modelxx.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   rate   R-squared:                       0.771
Model:                            OLS   Adj. R-squared:                  0.745
Method:                 Least Squares   F-statistic:                     30.22
Date:                Thu, 14 Oct 2021   Prob (F-statistic):           8.86e-09
Time:                        16:12:16   Log-Likelihood:                -43.614
No. Observations:                  31   AIC:                             95.23
Df Residuals:                      27   BIC:                             101.0
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.0543      2.747      3.297      0.003       3.419      14.690
len           -0.0853      0.027     -3.160      0.004      -0.141      -0.030
slim          -0.0935      0.046     -2.042      0.051      -0.187       0.000
acpt           0.1047      0.027      3.932      0.001       0.050       0.159
==============================================================================
Omnibus:                        0.336   Durbin-Watson:                   2.784
Prob(Omnibus):                  0.845   Jarque-Bera (JB):                0.403
Skew:                           0.220   Prob(JB):                        0.818
Kurtosis:                       2.656   Cond. No.                         839.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R^2 is 0.771

### Problem 5
Make a prediction if $y$ values using testing data set and the model in problem 4. Finally find testing set $R^2$ and compare with the training set $R^2$.
Comment on how they compare and what does it mean.

In [102]:
ypred = modelxx.predict(sm.add_constant(xtest))

In [103]:
ypred

8     3.817329
30    4.867455
33    3.584943
19    2.528074
35    4.888878
16    3.761718
34    2.777428
18    4.473905
dtype: float64

In [108]:
from sklearn.metrics import r2_score
r2_score(ytest, ypred)

-0.4338302155331446

The testing set R^2 has a negative correlation while the training set R^2 has a strong positive correlation. This means that the relationship of the variable in the testing set is negative but in the training set it is positive. 